Update floats.md

wheresaddie opened the pull request:
### Floating Point in Substrate explained

Thank you for making a PR to the knowledge base. Please make sure it follows these short guidelines:

- [ ] Are the audience and objective of the document clear?
- [ ] Is the writing:
  - **Clear**: All jargon defined, explained, or linked to. Best to avoid it.
  - **Precise**: No ambiguous meanings
  - **Concise**: Free of superfluous detail
- [ ] Does it follow our [style guide](/CONTRIBUTING.md#documentation-style)?
- [ ] If this is a new page, does the PR include the appropriate infrastructure, e.g. adding the
      page to a sidebar?
- [ ] Build the page. Does it render properly? E.g. no funny lists or formatting.
- [ ] Links should go to rustdocs or devhub articles, not source code.
- [ ] If this PR addresses an issue in the queue, have you referenced it in the description?

wheresaddie committed:
Update floats.md

### Floating Point in Substrate explained

apopiak commented on current/runtime/floats.md:
```diff
@@ -4,9 +4,11 @@ lang: en
+When developing for the Substrate runtime and doing calculations it is appealing to use floating point numbers for e.g. representing fractions. In Substrate it is important that each node arrives at the same deterministic result for consensus. This document covers why traditional floating point numbers are problematic in runtime and what alternatives exist.
 
-## Problems with Floating Point Numbers in the Runtime
+## Problems with Floating Point Numbers in Runtime
```
Curious why you opted for dropping the "the". AFAIK we use runtime generically and not as Named Thing (i.e. Runtime)? Oh or are you doing this because you use like "in context"? Like "We don't use non-deterministic operations in runtime [development]"?

wheresaddie replied:
yes-- using this in context- if you prefer generically calling it-- imo works both ways

apopiak replied:
I think I've only seen it used generically while working at Parity (which hasn't been that long ^^) and would personally make it explicit (i.e. use "runtime development") when using it as an activity.
But in the end I only care whether it's consistent (over our communication and docs). I could get used to your variant *shrug*

apopiak commented on current/runtime/floats.md:
```diff
@@ -4,9 +4,11 @@ lang: en
-## Problems with Floating Point Numbers in the Runtime
+## Problems with Floating Point Numbers in Runtime
+
+The standard way of handling floating points is dependent on the need for decimal places by using one of the in-built primitive types of handlingfix point arithmetic. Floating point numbers are not deterministic and fixed point computation is safe for Substrate since it represents all rationals as a fraction thus resolving a deterministic result. The two types which are used to handle large fixed ...
```
```suggestion
Floating point numbers are not deterministic and thus cannot be used to run a deterministic block production algorithm. The standard way of handling rational (or real) numbers depends on how many decimal places are needed and then choosing one of the in-built primitive types in Substrate providing fixed point arithmetic. Fixed point computation is safe for Substrate since it represents all rationals as a fraction thus resolving to a deterministic result. The two types which are used to handle large fixed point arithmetic are `Permill` and `Perbill`, you can refer to [safe math](https://substrate.dev/recipes/3-entrees/safemath.html) for further information and implementation.
```

apopiak replied:
A few comments:
+ AFAICT This paragraph does not seem to focus on the problems with floating point numbers, so you might want to tease those things apart?
+ This is a knowledge base article so I'm not sure we want to recommend `Permill` and `Perbill` straight away, but instead explain the general/transferable reasoning.
+ I think people (including myself) easily confuse and conflate floats with rational/real numbers when they are distinct. So I would want to avoid using "floats" in places that refer to real/rational numbers. About which words to use for describing this: IMO float (floating point number), rational (rational number) and fixed point number are all seperate concepts. I, personally, would try to always use them in a way that distinguishes them from each other and makes sure that they remain distinct in the context of this article:
  + rational numbers are a mathematical concept of numbers that can be represented by the quotient of 2 integers https://en.wikipedia.org/wiki/Rational_number
  + floats are an encoding (for real numbers, including rational numbers) with limited precision
  + fixed point numbers are a different encoding also with limited precision

wheresaddie replied:
got it -- 
focus on floating point number issues. 
explain general/transferable reasons vs 'permill' 
possibly define rational/real numbers when referring to floats

wheresaddie committed:
Update current/runtime/floats.md

going to add an additional explanation of general/transferable reasoning after 'Perbill'/ 'Permill' for next PR

Co-authored-by: Alexander Popiak <alexander.popiak@gmail.com>
